cgen: error on generic comptime selector with nested structs

### Describe the bug

Comptime selector can only handle one nested struct type.
Trying to register multiple generic types results in a C gen error about incompatible types. 

### Expected Behavior

Types are correctly resolved.

### Current Behavior
```
/tmp/v_501/ct_struct2.7639474865608158254.tmp.c:27593:58: warning: incompatible pointer types passing 'main__Child1 *' (aka 'struct main__Child1 *') to para
meter of type 'main__Child2 *' (aka 'struct main__Child2 *') [-Wincompatible-pointer-types]
                                res.child1 = main__decode_struct_T_main__Child2(val, &s_);
                                                                                     ^~~
/tmp/v_501/ct_struct2.7639474865608158254.tmp.c:3878:78: note: passing argument to parameter 'typ' here
main__Child2 main__decode_struct_T_main__Child2(toml__Any doc, main__Child2* typ);
                                                                             ^
/tmp/v_501/ct_struct2.7639474865608158254.tmp.c:27593:16: error: assigning to 'main__Child1' (aka 'struct main__Child1') from incompatible type 'main__Child
2' (aka 'struct main__Child2')
                                res.child1 = main__decode_struct_T_main__Child2(val, &s_);
                                           ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
7 warnings and 1 error generated.
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

```v
import toml

struct Parent {
	name   string
	child1 Child1
	child2 Child2
}

struct Child1 {
	name string
}

struct Child2 {
	age int
}

fn decode[T](toml_str string) !T {
	doc := toml.parse_text(toml_str)!
	
	return decode_struct(doc.to_any(), &T{})
}

// `T` param serves here as a workaround to allow to infer types of nested struct fields.
fn decode_struct[T](doc toml.Any, typ &T) T {
	mut res := T{}
	$for field in T.fields {
		val := doc.value(field.name)
		$if field.typ is string {
			res.$(field.name) = val.string()
		} $else $if field.typ is int {
			res.$(field.name) = val.int()
		} $else $if field.is_struct {
			typ_ := typ.$(field.name)
			res.$(field.name) = decode_struct(val, &typ_)
		}
	}
	return res
}

toml_str := 'name = "John"
child1 = { name = "abc" }
child2 = { age = 5 }'

dump(decode[Parent](toml_str)!)
```

Works with just 1 Child.

Alternatively, passing a mut struct instead of returning a result results in wrong values instead of a c error.

<details><summary><code>mut</code> sturct reproduction</summary>

```v
import toml

struct Parent {
	name   string
	child1 Child1
	child2 Child2
}

struct Child1 {
	name string
}

struct Child2 {
	age int
}

fn decode[T](toml_str string) !T {
	doc := toml.parse_text(toml_str)!
	mut t := T{}
	// passing type as param would atm allow to infer types of nested struct fields
	decode_struct(doc.to_any(), mut t)
	return t
}

fn decode_struct[T](doc toml.Any, mut typ T) {
	$for field in T.fields {
		val := doc.value(field.name)
		$if field.typ is string {
			typ.$(field.name) = val.string()
		} $else $if field.typ is int {
			typ.$(field.name) = val.int()
		} $else $if field.is_struct {
			mut typ_ := typ.$(field.name)
			decode_struct(val, mut typ_)
			typ.$(field.name) = typ_
		}
	}
}

toml_str := 'name = "John"
child1 = { name = "abc" }
child2 = { age = 5 }'

dump(decode[Parent](toml_str)!)
```

</details>


### V version

0.3.4

### Environment details (OS name and version, etc.)

linux,arch;amd64

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1595,10 +1595,10 @@ fn (mut c Checker) resolve_comptime_args(func ast.Fn, node_ ast.CallExpr, concre
 			} else {
 				func.params[offset + i]
 			}
-			k++
 			if !param.typ.has_flag(.generic) {
 				continue
 			}
+			k++
 			param_typ := param.typ
 			if call_arg.expr is ast.Ident {
 				if call_arg.expr.obj is ast.Var {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1186,10 +1186,10 @@ fn (mut g Gen) resolve_comptime_args(func ast.Fn, mut node_ ast.CallExpr, concre
 			} else {
 				func.params[offset + i]
 			}
-			k++
 			if !param.typ.has_flag(.generic) {
 				continue
 			}
+			k++
 			param_typ := param.typ
 			if mut call_arg.expr is ast.Ident {
 				if mut call_arg.expr.obj is ast.Var {

--- a/vlib/v/tests/generic_comptime_arg_test.v
+++ b/vlib/v/tests/generic_comptime_arg_test.v
@@ -1,0 +1,53 @@
+import toml
+
+struct Parent {
+	name   string
+	child1 Child1
+	child2 Child2
+}
+
+struct Child1 {
+	name string
+}
+
+struct Child2 {
+	age int
+}
+
+fn decode[T](toml_str string) !T {
+	doc := toml.parse_text(toml_str)!
+
+	return decode_struct(doc.to_any(), &T{})
+}
+
+// `T` param serves here as a workaround to allow to infer types of nested struct fields.
+fn decode_struct[T](doc toml.Any, typ &T) T {
+	mut res := T{}
+	$for field in T.fields {
+		val := doc.value(field.name)
+		$if field.typ is string {
+			res.$(field.name) = val.string()
+		} $else $if field.typ is int {
+			res.$(field.name) = val.int()
+		} $else $if field.is_struct {
+			typ_ := typ.$(field.name)
+			res.$(field.name) = decode_struct(val, &typ_)
+		}
+	}
+	return res
+}
+
+fn main() {
+	toml_str := 'name = "John"
+	child1 = { name = "abc" }
+	child2 = { age = 5 }'
+
+	a := dump(decode[Parent](toml_str)!)
+	assert a.name == 'John'
+	assert a.child1 == Child1{
+		name: 'abc'
+	}
+	assert a.child2 == Child2{
+		age: 5
+	}
+}

--- a/vlib/v/tests/generic_comptime_arg_test.v
+++ b/vlib/v/tests/generic_comptime_arg_test.v
@@ -37,7 +37,7 @@ fn decode_struct[T](doc toml.Any, typ &T) T {
 	return res
 }
 
-fn main() {
+fn test_main() {
 	toml_str := 'name = "John"
 	child1 = { name = "abc" }
 	child2 = { age = 5 }'

